test/realtime/resume.test.js

```diff
@@ -629,5 +622,121 @@ define(['shared_helper', 'async', 'chai'], function (helper, async, chai) {
+          const MSG_COUNT = 5;
+
+          const txRest = helper.AblyRest();
+          const rxRealtime = helper.AblyRealtime(mixin(realtimeOpts));
+
+          const channelNames = Array(10).fill().map(() => String(Math.random()));
+          const rxChannels = channelNames.map((name) => rxRealtime.channels.get(name));
+
+          let recoveryKey;
+          let rxRealtimeRecover;
+          let rxRecoverChannels;
+
+          async.series(
+            [
+              function (cb) {
+                async.parallel(rxChannels.map((channel) => function (cb) {
+                  channel.attach(cb);
+                }), cb);
+              },
+              function (cb) {
+                function sendAndAwait(_, cb) {
+                  async.parallel(channelNames.map((name) => function (cb) {
+                    const tx = txRest.channels.get(name);
+                    const rx = rxRealtime.channels.get(name);
+                    const event = String(Math.random());
+                    async.parallel(
+                      [
+                        function (cb) {
+                          tx.publish(event, null, cb);
+                        },
+                        function (cb) {
+                          rx.subscribe(event, function (msg) {
+                            rx.unsubscribe(event);
+                            cb();
+                          });
+                        },
+                      ],
+                      cb
+                    )
+                  }), cb);
+                }
+                // Publish messages on each channel.
+                async.timesSeries(MSG_COUNT, sendAndAwait, cb);
+              },
+              function (cb) {
+                recoveryKey = rxRealtime.connection.getRecoveryKey();
+                helper.becomeSuspended(rxRealtime, cb);
+              },
+              function (cb) {
+                function publishAll(_, cb) {
+                  async.parallel(channelNames.map((name) => function (cb) {
+                    const tx = txRest.channels.get(name);
+                    tx.publish('sentWhileDisconnected', null, cb);
+                  }), cb);
+                }
+                // Publish another MSG_COUNT messages on each channel (which are no
+                // longer received by rxRealtime).
+                async.timesSeries(MSG_COUNT, publishAll, cb);
+              },
+              function (cb) {
+                rxRealtimeRecover = helper.AblyRealtime({ recover: recoveryKey });
+                rxRecoverChannels = channelNames.map((name) => rxRealtimeRecover.channels.get(name));
+                cb();
+              },
+              function (cb) {
+                // Subscribe to the first channels before becoming connected
+                // (so when the transport is activated the channels serials
+                // should be set).
+                async.parallel(rxRecoverChannels.slice(0, 5).map((channel) => function (cb) {
```

Comment:
How do we make sure the connection is not yet in a `connected` state here?

Reply:
Comment says to subscribe to channels before becoming `connected`, the `attach` cannot happen without connection being `connected`